### PR TITLE
feat: add stateful model support for server-side conversation management

### LINKED ISSUE #1957: [FEATURE] Support previous_response_id for stateful multi-turn conversations in OpenAIResponsesModel
### Problem Statement

 The OpenAIResponsesModel currently resends the full conversation history on every turn, even though the Responses API natively
  supports server-side conversation state via previous_response_id. The docstring at the top of openai_responses.py acknowledges
  this:

> "The Responses API can maintain conversation state server-side through 'previous_response_id'... Note: This implementation currently only implements the stateless approach."

  For agentic applications with multi-turn conversations (10+ turns with tool calls), this means:
  - Token costs scale linearly with conversation length — turn 10 resends all previous turns
  - Latency increases as the input payload grows
  - Context window pressure — long conversations hit limits faster, not because of new content, but because of repeated history

  This affects both OpenAI and xAI Responses API users, since both support previous_response_id.

### Proposed Solution

1. After each successful response in stream(), capture and store response.id from the completed response event.
  2. On subsequent calls, if a previous_response_id is available, pass it in the request instead of the full message history. Only
   send the new user message(s) and tool results in input.
  3. Fall back to the current stateless approach (full history) if:
    - No previous response ID exists (first turn)
    - The stored response has expired (30-day server retention)
    - The API returns an error indicating the previous response is invalid
  4. Expose a configuration option to enable/disable this behavior (e.g., stateful=True in config), defaulting to disabled for
  backward compatibility.

  The key changes would be in:
  - stream() — capture response.id from response.completed event
  - _format_request() — conditionally pass previous_response_id instead of full history
  - State management — store the last response ID (could be returned as metadata alongside usage stats)

### Use Case

  We run a property management AI agent on Bedrock AgentCore using Strands. Each session is a multi-turn conversation where the PM
   asks for analysis, creates action plans, drafts emails, and executes tasks. A typical session is 10-20 turns with heavy tool
  use (each turn may involve 2-5 tool calls).

  Today, turn 15 of a conversation resends all 14 previous turns plus their tool call/result pairs. With previous_response_id,
  turn 15 would only send the new user message — the server already has the rest.

  This would help with:
  - Cost reduction — estimated 40-60% input token savings for typical multi-turn sessions
  - Faster responses — less data to transmit and process per turn
  - Longer conversations — more room in the context window for actual content instead of repeated history

### Alternatives Solutions

Application-level caching/summarization - can get messy

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by pgrayy**
## Motivation

The OpenAI Responses API supports server-side conversation state management, where the server tracks context across turns so the SDK sends only the latest message instead of the full history each time. This PR adds support for this capability as a model-agnostic feature through a new `stateful` config on the `Model` base class.

Previously, models were always stateless: the full conversation history was sent with every request. For providers that can manage conversation state server-side, this is unnecessary overhead. The `stateful` property lets models opt into server-managed conversations, with the framework handling message cleanup and conversation manager coordination automatically.

Resolves: #1957

Documentation will follow in a separate PR.

### Enabling server-side conversation management

A new `stateful` config option on `OpenAIResponsesModel` enables server-side conversation state.

```python
from strands import Agent
from strands.models.openai_responses import OpenAIResponsesModel

model = OpenAIResponsesModel(
    model_id="gpt-4o",
    stateful=True,
)
agent = Agent(model=model, system_prompt="You are a helpful assistant.")

agent("My name is Alice.")
# agent.messages is empty (server owns history)

agent("What is my name?")
# Server remembers "Alice" via previous_response_id
```

When `stateful=True`, the agent automatically uses `NullConversationManager` and raises `ValueError` if a conversation manager is explicitly passed.

### Model.stateful property

The `Model` base class now exposes a `stateful` property (defaults to `False`). Model providers that support server-side state override this. `OpenAIResponsesModel` derives it from the `stateful` config option.

### ModelPlugin

A new `_ModelPlugin` (derived from `Plugin`) handles model lifecycle hooks. It registers an `AfterInvocationEvent` callback that checks `Model.stateful` directly and clears messages when the model is operating in stateful mode.

### model_state passed to Model.stream

`stream_messages` passes `model_state` (the runtime state dict) to `Model.stream` via kwargs. This gives model providers access to runtime state like `response_id` for conversation chaining. The model owns both reading and writing this state: `OpenAIResponsesModel` reads `response_id` to set `previous_response_id` in requests, and captures the new `response_id` from the `response.created` streaming event. The `previous_response_id` is only sent when `Model.stateful` is true.

### Session persistence

`_model_state` is persisted in `SessionAgent._internal_state`, so server-side conversations survive process restarts when using a session manager. The `repository_session_manager` checks `Model.stateful` directly to decide whether to skip message restoration.

### Multi-agent support

Graph and Swarm nodes capture and restore `_model_state` alongside messages and state during resets and interrupt save/restore cycles.

## Related Issues

#1957

## Type of Change

New feature

- [x] I ran `hatch run prepare`

## Follow-ups

- **Reasoning content in request formatting**: `OpenAIResponsesModel._format_request_message_content` does not handle `reasoningContent` blocks. Stateful mode sidesteps this since messages are cleared.
- **Built-in tool support**: The Responses API supports server-side tools (web search, code interpreter, file search). These are not yet handled in the streaming event parser.
- **Factory pattern for http_client**: The OpenAI SDK client is recreated each call, closing any provided `http_client`. A factory pattern would avoid the non-closing client workaround. See #1036.

## Checklist
- [x] I have read the CONTRIBUTING document
- [x] I have added any necessary tests that prove my fix is effective or my feature works
- [ ] I have updated the documentation accordingly
- [x] I have added an appropriate example to the documentation to outline the feature, or no new docs are needed
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published

----

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/strands/models/openai_responses.py [hunk 1]
```diff
@@ -1,18 +1,8 @@
 """OpenAI model provider using the Responses API.
 
-The Responses API is OpenAI's newer API that differs from the Chat Completions API in several key ways:
+Note: Built-in tools (web search, code interpreter, file search) are not yet supported.
```
Comment (pgrayy) [inline src/strands/models/openai_responses.py:3]: Going to address this in a follow up PR.

File: src/strands/models/openai_responses.py [hunk 2]
```diff
@@ -1,18 +1,8 @@
...
 
-1. The Responses API can maintain conversation state server-side through "previous_response_id",
-   while Chat Completions is stateless and requires sending full conversation history each time.
-   Note: This implementation currently only implements the stateless approach.
-
-2. Responses API uses "input" (list of items) instead of "messages", and system
-   prompts are passed as "instructions" rather than a system role message.
-
-3. Responses API supports built-in tools (web search, code interpreter, file search)
-   Note: These are not yet implemented in this provider.
```
Comment (pgrayy) [inline src/strands/models/openai_responses.py:13]: Removed all this since we link the openai documentation.

File: tests_integ/models/test_model_mantle.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+            headers=dict(request.headers),
+            data=request.content,
+        )
+        self.signer.add_auth(aws_request)
+        for key, value in aws_request.headers.items():
+            request.headers[key] = value
+        yield request
+
+
+class _NonClosingAsyncClient(httpx.AsyncClient):
```
Comment (pgrayy) [inline tests_integ/models/test_model_mantle.py:34]: Going to simplify in a follow up PR. If we have OpenAIResponsesModel accept a factory function for the http client, we won't need to do this (https://github.com/strands-agents/sdk-python/issues/1036)

File: src/strands/agent/conversation_manager/conversation_manager.py [hunk 4]
```diff
@@ -60,7 +67,21 @@ def register_hooks(self, registry: HookRegistry, **kwargs: Any) -> None:
...
+
+        Performs the following:
+        - Clears messages when the model is managing conversation state server-side.
+        """
+        model_state = getattr(event.agent, "_model_state", {})
+        if model_state.get("stored"):
+            event.agent.messages.clear()
+            logger.debug(
+                "response_id=<%s> | cleared messages for server-managed conversation",
+                model_state["response_id"],
```
[the remark this answers is not in the document]
Comment (pgrayy) [reply src/strands/agent/conversation_manager/conversation_manager.py]: As you stated, should always be present.

File: src/strands/models/openai_responses.py [hunk 5]
```diff
@@ -400,13 +399,17 @@ def _format_request(
                 format.
         """
         input_items = self._format_request_messages(messages)
-        request = {
+        request: dict[str, Any] = {
             "model": self.config["model_id"],
             "input": input_items,
             "stream": True,
+            "store": False,
```
Comment (pgrayy) [inline src/strands/models/openai_responses.py]: Responses are stored by default ([docs](https://developers.openai.com/api/docs/guides/migrate-to-responses)). This is not required however when using Strands because we manage the messages locally by default. Consequently, storing by default is wasteful.

File: src/strands/agent/agent.py [hunk 6]
```diff
@@ -282,6 +282,9 @@ def __init__(
 
         self._interrupt_state = _InterruptState()
 
+        # Runtime state for model providers (e.g., server-side response ids)
+        self._model_state: dict[str, Any] = {}
```
Comment (pgrayy) [inline src/strands/agent/agent.py:299]: Quick note, we don't accept the response id in Agent init or stream right now. This means you can only connect to stored conversations started using Strands. I think this is okay to start and what users would expect anyway. Accepting response ids would create a foot gun where users may accidentally setup multiple agents operating on the same conversation, thus creating conflict.

File: src/strands/agent/conversation_manager/conversation_manager.py [hunk 7]
```diff
@@ -60,7 +67,21 @@ def register_hooks(self, registry: HookRegistry, **kwargs: Any) -> None:
                 registry.add_callback(SomeEvent, self.on_some_event)
             ```
         """
-        pass
+        registry.add_callback(AfterInvocationEvent, self._on_after_invocation)
+
+    def _on_after_invocation(self, event: AfterInvocationEvent) -> None:
```
Comment (zastrowm) [inline src/strands/agent/conversation_manager/conversation_manager.py]: This seems a little odd; this implies the NoopConversationManager is going to start doing this as well?

It's debatable if every conversation-manager should have this logic; I would think for the majority of use cases you wouldn't' want a conversation manager with these new models, right?

File: src/strands/models/openai_responses.py [hunk 8]
```diff
@@ -180,6 +170,7 @@ async def stream(
         system_prompt: str | None = None,
         *,
         tool_choice: ToolChoice | None = None,
+        response_id: str | None = None,
```
Comment (zastrowm) [inline src/strands/models/openai_responses.py]: What's the reasoning of not passing in model-state?  I like that more than special casing response_id

File: src/strands/multiagent/graph.py [hunk 9]
```diff
@@ -192,6 +196,9 @@ def reset_executor_state(self) -> None:
         if hasattr(self.executor, "state"):
             self.executor.state = AgentState(self._initial_state.get())
 
+        if hasattr(self.executor, "_model_state"):
+            self.executor._model_state = copy.deepcopy(self._initial_model_state)
```
Comment (zastrowm) [inline src/strands/multiagent/graph.py:200]: What happens when there are backend messages stored and we have a request-id locally?  Does that imply that a graph would overwrite backend data on every call?  Would they get an exception instead?

File: src/strands/agent/conversation_manager/conversation_manager.py (hunk 7 above)
Comment (mkmeral) [inline src/strands/agent/conversation_manager/conversation_manager.py]: will this only apply to agents with conversation manager? why? 

I am a bit confused on DevX here, because I'd expect the customers who use Responses API not to actually use conversation manager (because they'd be no-op)

File: src/strands/types/streaming.py [hunk 10]
```diff
@@ -163,11 +163,15 @@ class MetadataEvent(TypedDict, total=False):
         metrics: Performance metrics related to the model invocation.
         trace: Trace information for debugging and monitoring.
         usage: Resource usage information for the model invocation.
+        responseId: Server-generated response identifier.
+        stored: Whether the response was stored server-side for conversation chaining.
```
Comment (zastrowm) [inline src/strands/types/streaming.py]: nits:

 - should it be response_id? (I don't know TBH)
 - do we have a better name for `stored`?  It's not clear from the naming what is `stored`

File: src/strands/agent/conversation_manager/conversation_manager.py [hunk 11]
```diff
@@ -1,14 +1,18 @@
...
 from ...hooks.registry import HookProvider, HookRegistry
 from ...types.content import Message
 
 if TYPE_CHECKING:
     from ...agent.agent import Agent
 
+logger = logging.getLogger(__name__)
+
 
 class ConversationManager(ABC, HookProvider):
```
Comment (zastrowm) [inline src/strands/agent/conversation_manager/conversation_manager.py:13]: Does using a ConversationManager with stateful backends even make sense?  I wonder if it should be an error TBH?

Or alternatively, if the default conversation manager for a stateful model is the Noop provider and then throw if a different one is passed

File: src/strands/models/openai_responses.py (hunk 2 above)
Comment (mkmeral) [reply src/strands/models/openai_responses.py:13]: nit (and i'm not sure if it applies here): but let's leave info in the files + docstrings. that's the source of truth that agents + IDEs read

File: src/strands/models/openai_responses.py (hunk 5 above)
Comment (mkmeral) [reply src/strands/models/openai_responses.py]: is `store` the API param? I wonder if we can move it to top level in agent constructor, so it's a bit more integrated use case (bc it is, we default to a different behavior)

```python
model = OpenAIResponsesModel(
    model_id="gpt-4o",
    params={"store": True},
)
```

instead 
```python
model = OpenAIResponsesModel(
    model_id="gpt-4o",
    store=True,
)
```

**Review by agent-of-mkmeral — COMMENTED** [review]
**TL;DR**: FAIL — 1 bug found (session state corruption via missing deep copy). 43/44 adversarial tests pass.

<details>
<summary>Adversarial Testing Report</summary>

**Scope**: Server-side conversation state management — model_state lifecycle, ConversationManager hook, streaming metadata capture, session persistence, OpenAI Responses model formatting, Graph/Swarm multi-agent state, backward compatibility.
**Tests written**: 44
**Tests passing**: 43
**Tests failing (findings)**: 1

| # | Category | Severity | Description |
|---|----------|----------|-------------|
| 1 | Bug | Medium | `SessionAgent.from_agent()` stores `_model_state` by reference — session state corrupted by subsequent agent mutations |

<details>
<summary>Finding 1 — Bug (Medium): SessionAgent.from_agent() missing deep copy of _model_state</summary>

**Root cause**: `src/strands/types/session.py` line 137 stores `agent._model_state` directly instead of `copy.deepcopy(agent._model_state)`. This means the saved session snapshot shares the same dict object as the live agent.

**Reproduction**:
```python
from strands.types.session import SessionAgent
from strands.agent.agent import Agent

agent = Agent(model=model, callback_handler=None)
agent._model_state = {"response_id": "resp_1", "stored": True}

session = SessionAgent.from_agent(agent)
assert session._internal_state["model_state"]["response_id"] == "resp_1"  # ✅

agent._model_state["response_id"] = "MUTATED"
assert session._internal_state["model_state"]["response_id"] == "resp_1"  # ❌ FAILS — returns "MUTATED"
```

**Observed behavior**: After `from_agent()`, mutating `agent._model_state` also mutates the saved `SessionAgent._internal_state["model_state"]` because they are the same dict object.

**Expected behavior**: `from_agent()` should produce an isolated snapshot. Mutations to the live agent after saving should not affect the saved session.

**Fix**:
```python
# src/strands/types/session.py, line ~137
_internal_state={
    "interrupt_state": agent._interrupt_state.to_dict(),
    "model_state": copy.deepcopy(agent._model_state),  # was: agent._model_state
},
```

Note: `interrupt_state` uses `.to_dict()` which creates a new dict, so it doesn't have this problem. Only `_model_state` is stored by reference.

</details>

<details>
<summary>Edge Cases Validated (not bugs, but worth noting)</summary>

**Defensive coding gaps** (not blocking, but hardening opportunities):
- `ConversationManager._on_after_invocation()` accesses `model_state["response_id"]` with direct dict lookup. If `stored=True` but `response_id` is missing, this raises `KeyError`. The reviewer already flagged this and the author declined `.get()`. The test confirms the behavior — it's a conscious tradeoff.
- `process_stream()` accesses `chunk["metadata"]["responseId"]` with direct dict lookup when `stored=True`. If a model emits `stored` without `responseId`, this is a `KeyError`. Acceptable since the contract expects both together.

</details>

<details>
<summary>What Survived (43 tests passed)</summary>

- ✅ **Model state lifecycle**: Initialization, multi-turn updates, response_id passthrough
- ✅ **ConversationManager hook**: All edge cases (missing state, explicit False, None, empty, derived classes)
- ✅ **Streaming metadata**: Stored/not-stored paths, missing field handling
- ✅ **Session persistence**: Save/restore, old session backward compatibility
- ✅ **OpenAI Responses model**: store default, params override, previous_response_id mapping, format_chunk metadata
- ✅ **MetadataEvent type**: New fields accepted, old format still works
- ✅ **Graph nodes**: Deep copy isolation, reset behavior, empty state
- ✅ **Swarm nodes**: Deep copy isolation, reset with/without context, old context backward compatibility
- ✅ **Contract verification**: All PR claims validated (messages cleared, response_id chaining, all managers get hook)
- ✅ **Backward compatibility**: Stateless agents unaffected, serialization unchanged

</details>

</details>

---
🤖 *AI agent response. [Strands Agents](https://github.com/strands-agents). Feedback welcome!*

File: src/strands/models/openai_responses.py (hunk 8 above)
Comment (pgrayy) [reply src/strands/models/openai_responses.py]: Model state is something specific to the Agent. It is its model orchestration runtime data. The model's themselves should have to care and instead can receive only what they need. That was my thinking at least.

File: src/strands/agent/conversation_manager/conversation_manager.py (hunk 7 above)
Comment (pgrayy) [reply src/strands/agent/conversation_manager/conversation_manager.py]: In the majority of cases yeah you probably would not care about what conversation manager is used. It may come up though if for example you needed special handling of your user messages like the tool results. Or if your agent looped many times on a single user request, summarization could be handy. Admittedly, that would have to be a lot of loops.

With all that said, I think it still makes sense because you always need your local messages managed when using a stateful model. Centralizing here seemed like a natural fit. Even for the NoopConverationManager. It isn't the implementation itself doing it but the ConversationManager.

File: src/strands/multiagent/graph.py (hunk 9 above)
Comment (pgrayy) [reply src/strands/multiagent/graph.py:200]: So the response_id works like a linked list. If we keep passing in the same response id, the conversation server side will branch off many times from the same message rather than appending linearly.

File: src/strands/types/streaming.py (hunk 10 above)
Comment (pgrayy) [reply src/strands/types/streaming.py]: Given the context, I thought it made sense. This is metadata for a streaming response from the model. I could try to workshop other names though.

File: src/strands/agent/conversation_manager/conversation_manager.py (hunk 11 above)
Comment (pgrayy) [reply src/strands/agent/conversation_manager/conversation_manager.py:13]: As mentioned above, using conversation managers could still come into play. For example, maybe someone uses a conversation manager to help properly size large tool results.

File: src/strands/agent/conversation_manager/conversation_manager.py (hunk 7 above)
Comment (pgrayy) [reply src/strands/agent/conversation_manager/conversation_manager.py]: As mentioned above, using conversation managers could still come into play. For example, maybe someone uses a conversation manager to help properly size large tool results.

File: src/strands/models/openai_responses.py (hunk 2 above)
Comment (pgrayy) [reply src/strands/models/openai_responses.py:13]: - Comment 1: This one is outdated since stateful mode is now supported and this is documented more general on Model.stateful.
- Comment 2: This is an implementation detail and I don't see it as particularly helpful anyway for a model.
- Comment 3: I left a note that built-in tools are still unsupported. I have a follow up PR that adds support for built-in tools and creates clear documentation for them up here in the top level docstring.

Generally I agree with the point you raise.

File: src/strands/models/openai_responses.py (hunk 5 above)
Comment (pgrayy) [reply src/strands/models/openai_responses.py]: Addressed. Added a top level `stateful` config param.

File: src/strands/agent/conversation_manager/conversation_manager.py (hunk 7 above)
Comment (pgrayy) [reply src/strands/agent/conversation_manager/conversation_manager.py]: Addressed. Moved logic into an internal _ModelPlugin.

File: src/strands/models/openai_responses.py (hunk 8 above)
Comment (pgrayy) [reply src/strands/models/openai_responses.py]: Addressed. Passing in model_state now.

File: src/strands/types/streaming.py (hunk 10 above)
Comment (pgrayy) [reply src/strands/types/streaming.py]: The params are responseId and stateful now. I feel from the context, the values names make sense. We also have the docstrings.

File: src/strands/agent/conversation_manager/conversation_manager.py (hunk 11 above)
Comment (pgrayy) [reply src/strands/agent/conversation_manager/conversation_manager.py:13]: Addressed. Now raise an error if assigning a conversation manager when running a model in stateful mode.

File: src/strands/event_loop/streaming.py [hunk 12]
```diff
@@ -448,6 +450,9 @@ async def process_stream(
                 int(1000 * (first_byte_time - start_time)) if (start_time and first_byte_time) else None
             )
             usage, metrics = extract_usage_metrics(chunk["metadata"], time_to_first_byte_ms)
+            if model_state is not None and chunk["metadata"].get("stateful"):
+                model_state["response_id"] = chunk["metadata"]["responseId"]
+                model_state["stateful"] = True
```
Comment (zastrowm) [inline src/strands/event_loop/streaming.py]: Any reason not to to put this into the model?

File: src/strands/models/model.py [hunk 13]
```diff
@@ -115,3 +129,35 @@ def stream(
...
+
+    @staticmethod
+    def _on_after_invocation(event: AfterInvocationEvent) -> None:
+        """Handle post-invocation model management tasks.
+
+        Performs the following:
+        - Clears messages when the model is managing conversation state server-side.
+        """
+        model_state = getattr(event.agent, "_model_state", {})
+        if model_state.get("stateful"):
```
Comment (zastrowm) [inline src/strands/models/model.py]: This should be `event.agent.model.stateful`, no?

File: src/strands/models/openai_responses.py [hunk 14]
```diff
@@ -291,8 +297,10 @@ async def stream(
 
                         elif event.type == "response.completed":
                             # Response complete
-                            if hasattr(event, "response") and hasattr(event.response, "usage"):
-                                final_usage = event.response.usage
+                            if hasattr(event, "response"):
+                                metadata["response_id"] = getattr(event.response, "id", None)
```
Comment (zastrowm) [inline src/strands/models/openai_responses.py]: This can't be done in the model?  E.g. we still have to special casing the id stuff in the agent loop?

File: src/strands/session/repository_session_manager.py [hunk 15]
```diff
@@ -220,11 +224,21 @@ def initialize(self, agent: "Agent", **kwargs: Any) -> None:
             if len(session_messages) > 0:
                 self._latest_agent_message[agent.agent_id] = session_messages[-1]
 
-            # Restore the agents messages array including the optional prepend messages
-            agent.messages = prepend_messages + [session_message.to_message() for session_message in session_messages]
-
-            # Fix broken session histories: https://github.com/strands-agents/sdk-python/issues/859
-            agent.messages = self._fix_broken_tool_use(agent.messages)
+            # Skip restoring messages when conversation is managed server-side
+            if agent._model_state.get("stateful"):
```
Comment (zastrowm) [inline src/strands/session/repository_session_manager.py]: Why not make this `agent.model.stateful`?

File: src/strands/event_loop/streaming.py (hunk 12 above)
Comment (pgrayy) [reply src/strands/event_loop/streaming.py]: Fixed. Removed stateful from model_state and Metadata. Now rely on Model.stateful.

File: src/strands/models/model.py (hunk 13 above)
Comment (pgrayy) [reply src/strands/models/model.py]: Fixed. Removed stateful from model_state and Metadata. Now rely on Model.stateful.

File: src/strands/session/repository_session_manager.py (hunk 15 above)
Comment (pgrayy) [reply src/strands/session/repository_session_manager.py]: Fixed. Removed stateful from model_state and Metadata. Now rely on Model.stateful.

File: src/strands/models/openai_responses.py (hunk 14 above)
Comment (pgrayy) [reply src/strands/models/openai_responses.py]: We don't have any special casing or conditioning on response_id in the agent loop but we do need to store it in model_state so that we can pass it in on the next turn (assigned to previous_response_id in the openai request). If we didn't have to worry about session management, we could keep this value in memory as an instance level field.

Side note, another alternative to yielding inside metadata is to just have stream here update model_state directly. That is an option now since we pass in model_state instead of response_id.

Comment (pgrayy) [reply src/strands/models/openai_responses.py]: Updated to have stream edit model state rather than send response id in the Metadata block.

**Review by zastrowm — APPROVED** [review]
